browser/net/brave_site_hacks_network_delegate_helper.cc

```diff
@@ -73,11 +74,35 @@ DECLARE_LAZY_MATCHER(tracker_appended_matcher,
 
 #undef DECLARE_LAZY_MATCHER
 
-void ApplyPotentialQueryStringFilter(const GURL& request_url,
-                                     std::string* new_url_spec) {
-  DCHECK(new_url_spec);
+void ApplyPotentialQueryStringFilter(std::shared_ptr<BraveRequestInfo> ctx) {
   SCOPED_UMA_HISTOGRAM_TIMER("Brave.SiteHacks.QueryFilter");
-  std::string new_query = request_url.query();
+
+  if (!ctx->initiator_url.is_valid()) {
+    // Direct navigations (e.g. omnibar, bookmarks) are exempted.
+    return;
+  }
+
+  if (ctx->redirect_source.is_valid()) {
+    if (ctx->internal_redirect) {
+      // Ignore internal redirects since we trigger them.
+      return;
+    }
+
+    if (net::registry_controlled_domains::SameDomainOrHost(
+            ctx->redirect_source, ctx->request_url,
+            net::registry_controlled_domains::INCLUDE_PRIVATE_REGISTRIES)) {
+      // Same-site redirects are exempted.
+      return;
+    }
+  } else if (net::registry_controlled_domains::SameDomainOrHost(
```

Comment:
maybe we also can shorten the code by chosing something like `source_url` conditionally?
```
if (...) {source_url = ctx->redirect_source;}
else if (...) {source_url = ctx->initiator_url;}

if (net::registry_controlled_domains::SameDomainOrHost(source_url, ctx->request_url, ...) {
  return;
}
```